src/redisearch_rs/trie_bencher/src/lib.rs

```diff
@@ -28,12 +29,37 @@ pub fn str2u8(input: &str) -> Vec<u8> {
     c_string.as_bytes().iter().copied().collect()
 }
 
-/// Convert a string to a pointer to a `c_char` array.
-/// It also returns the length of the string.
+/// Stores an input string for the C trie map.
 ///
-/// This is the expected input shape for insertions and retrievals on [`CTrieMap`].
-pub fn str2c_input(input: &str) -> (*mut c_char, u16) {
-    let converted = CString::new(input).expect("CString conversion failed");
-    let len: u16 = converted.as_bytes().len().try_into().unwrap();
-    (converted.into_raw(), len)
+/// The String isn't null terminated therefore we store the length and
+/// implements the `Drop` trait to free the memory when the struct goes out of scope.
+///
+/// This encapsulates the expected input shape for insertions and retrievals on [`CTrieMap`].
+pub struct StrCInput {
+    pub cstr_ptr: *mut c_char,
+    pub c_len: u16,
+}
+
+impl StrCInput {
+    pub fn new(input: &str) -> Self {
+        let converted = CString::new(input).expect("CString conversion failed");
+        let len: u16 = converted.as_bytes().len().try_into().unwrap();
+        Self {
+            cstr_ptr: converted.into_raw() as *mut c_char,
+            c_len: len,
+        }
+    }
+}
+
+impl Drop for StrCInput {
+    fn drop(&mut self) {
+        // free the memory allocated for the C string
+        unsafe {
+            dealloc(
+                self.cstr_ptr as *mut u8,
+                Layout::from_size_align(self.c_len as usize, std::mem::align_of::<usize>())
+                    .unwrap(),
+            )
+        }
+    }
 }
```

Comment:
Wouldn't it be simpler to define an extension method on `CString`, named `as_trie_input` which returns the underlying pointer and the correct length?

The implementation would desugar to

```rust
let view = self.as_c_str();
let len = view.count_bytes();
let ptr = view.as_ptr();
(ptr, len)
```
```